File: app/src/main/java/bigbigdw/joaradw/fragment_new/NewTab.java
package bigbigdw.joaradw.fragment_new;

import android.content.res.AssetManager;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;

import org.json.JSONObject;

import java.util.ArrayList;

import bigbigdw.joaradw.BookPagination;
import bigbigdw.joaradw.JOARADW;
import bigbigdw.joaradw.base.BookBaseFragment;
import bigbigdw.joaradw.etc.API;
import bigbigdw.joaradw.etc.BookList;
import bigbigdw.joaradw.main.MainBookListAdapterC;
import bigbigdw.joaradw.main.MainBookDataJSON;
import bigbigdw.joaradw.main.TabViewModel;
import bigbigdw.joaradw.main.MainBookListData;
import bigbigdw.joaradw.R;

public class NewTab extends BookBaseFragment {
    private MainBookListAdapterC adapter;
    private final ArrayList<MainBookListData> items = new ArrayList<>();
    private static final String ARG_SECTION_NUMBER = "section_number";
    private TabViewModel tabviewmodel;
    LinearLayout wrap;
    LinearLayout cover;
    LinearLayout blank;
    String store = "";
    String userToken;
    String etc = "";
    String classes = "&class=";
    JSONObject getuserinfo;
    LinearLayoutManager linearLayoutManager;
    RecyclerView recyclerView;
    RequestQueue queue;

    public static NewTab newInstance(int index) {
        NewTab fragment = new NewTab();
        Bundle bundle = new Bundle();
        bundle.putInt(ARG_SECTION_NUMBER, index);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        tabviewmodel = new ViewModelProvider(this).get(TabViewModel.class);
        int index = 1;
        if (getArguments() != null) {
            index = getArguments().getInt(ARG_SECTION_NUMBER);
        }
        tabviewmodel.setIndex(index);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_new_tab, container, false);
        queue = Volley.newRequestQueue(requireActivity());
        recyclerView  = root.findViewById(R.id.Main_NewBookList);
        wrap = root.findViewById(R.id.TabWrap);
        cover = root.findViewById(R.id.LoadingLayout);
        blank = root.findViewById(R.id.BlankLayout);
        adapter = new MainBookListAdapterC(items);
        linearLayoutManager = new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false);

        checkToken();
        JOARADW app = (JOARADW) requireActivity().getApplicationContext();
        userToken = app.getToken();

        AssetManager assetManager = requireActivity().getAssets();

        tabviewmodel.getText().observe(getViewLifecycleOwner(), tabNum -> {
            switch (tabNum) {
                case "TAB1":
                    store = "";
                    newBookList();
                    break;
                case "TAB2":
                    newBookListJSON(root, assetManager, "Main_Tab_Best_77FES.json");
                    break;
                case "TAB3":
                    newBookListJSON(root, assetManager, "Main_KidamuBookList.json");
                    break;
                case "TAB4":
                    newBookListJSON(root, assetManager, "Main_PromisedBookList.json");
                    break;
                case "TAB5":
                    store = "nobless";
                    newBookList();
                    break;
                case "TAB6":
                    store = "premium";
                    newBookList();
                    break;
                case "TAB7":
                    store = "series";
                    newBookList();
                    break;
                case "TAB8":
                    store = "finish";
                    newBookList();
                    break;
                default:
                    classes = "&class=short";
                    newBookList();
                    break;
            }
        });

        return root;
    }
    public void newBookList(){
        etc = "&store=" + store + "&orderby=redate&offset=25&page=" + 1 + "&token=" + userToken + classes;
        BookPagination.populateData(API.BOOK_LIST_JOA, etc, queue, wrap, items, cover, blank);
        BookList.initAdapterC(recyclerView, linearLayoutManager, adapter);
        BookPagination.scrollListener(API.BOOK_LIST_JOA, queue, wrap, items, adapter, recyclerView, etc);

        adapter.setOnItemClickListener((v, position, value) -> {
            MainBookListData item = adapter.getItem(position);
            adapterListener(item, value, queue);
        });
    }


    public void newBookListJSON(View root, AssetManager assetManager, String bookType){
        wrap.setVisibility(View.VISIBLE);
        cover.setVisibility(View.GONE);
        blank.setVisibility(View.GONE);

        RecyclerView recyclerViewJSON = root.findViewById(R.id.Main_NewBookList);
        LinearLayoutManager linearLayoutManagerJSON = new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL,false);
        recyclerViewJSON.setLayoutManager(linearLayoutManagerJSON);
        recyclerViewJSON.setAdapter(adapter);
        adapter.setItems(new MainBookDataJSON().getData(assetManager, bookType));
        adapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/bigbigdw/joaradw/fragment_new/NewTab.kt
package bigbigdw.joaradw.fragment_new

import android.content.Intent
import bigbigdw.joaradw.base.BookBaseFragment
import bigbigdw.joaradw.main.OLD_MainBookListAdapterC
import bigbigdw.joaradw.main.OLD_MainBookListData
import bigbigdw.joaradw.main.TabViewModel
import android.widget.LinearLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import android.view.LayoutInflater
import android.view.ViewGroup
import bigbigdw.joaradw.R
import bigbigdw.joaradw.JOARADW
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import bigbigdw.joaradw.book.*
import bigbigdw.joaradw.book_detail.BookDetailCover
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.ArrayList

class NewTab : Fragment() {
    private var adapter: AdapterBookListC? = null
    private val items = ArrayList<BookListDataC?>()
    private var tabviewmodel: TabViewModel? = null

    var wrap: LinearLayout? = null
    var cover: LinearLayout? = null
    var blank: LinearLayout? = null
    var token: String? = null
    var etc = ""
    var store = ""
    var linearLayoutManager: LinearLayoutManager? = null
    var recyclerView: RecyclerView? = null
    var page = 1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        tabviewmodel = ViewModelProvider(this).get(TabViewModel::class.java)
        var index = 1
        if (arguments != null) {
            index = requireArguments().getInt(ARG_SECTION_NUMBER)
        }
        tabviewmodel!!.setIndex(index)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        val root = inflater.inflate(R.layout.fragment_new_tab, container, false)
        val titleValues =
            requireContext()!!.getSharedPreferences("MAIN_MENU", AppCompatActivity.MODE_PRIVATE)
                .getString("NEW_TITLE", "")!!.replace("[", "").replace("]", "")
        val titleList: List<String> = titleValues!!.split(',').toList()

        token = requireContext()!!.getSharedPreferences("LOGIN", AppCompatActivity.MODE_PRIVATE)
            .getString("TOKEN", "")

        recyclerView = root.findViewById(R.id.Main_NewBookList)
        wrap = root.findViewById(R.id.TabWrap)
        cover = root.findViewById(R.id.LoadingLayout)
        blank = root.findViewById(R.id.BlankLayout)

        adapter = AdapterBookListC(requireContext(), items)
        linearLayoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)

        tabviewmodel!!.text.observe(viewLifecycleOwner, { tabNum: String? ->
            when (titleList[tabNum!!.replace("TAB", "").replace(" ", "").toInt()]) {
                "전체" -> {
                    store = ""
                    newBookList(page)
                }
                " 77FES" -> {
                    store = "festival"
                    newBookList(page)
                }
                " BORN" -> {
                    store = "joaraborn"
                    newBookList(page)
                }
                " 노블성실" -> {
                    store = "promise"
                    newBookList(page)
                }
                " 노블레스" -> {
                    store = "nobless"
                    newBookList(page)
                }
                " 노블X프리" -> {
                    store = "noblepre"
                    newBookList(page)
                }
                " 프리미엄" -> {
                    store = "premium"
                    newBookList(page)
                }
                " 무료" -> {
                    store = "series"
                    newBookList(page)
                }
                " 완결" -> {
                    store = "finish"
                    newBookList(page)
                }
                else -> {
                    store = "gidamu_event"
                    newBookList(page)
                }
            }
        })

        recyclerView!!.addOnScrollListener(recyclerViewScroll)

        return root
    }

    private fun newBookList(page : Int?) {
        RetrofitBookList.getNewBook(token, store, page)!!.enqueue(object : Callback<BookListResultC?> {
            override fun onResponse(
                call: Call<BookListResultC?>,
                response: Response<BookListResultC?>
            ) {

                if (response.isSuccessful) {
                    cover!!.visibility = View.GONE
                    blank!!.visibility = View.GONE
                    response.body()?.let { it ->
                        val books = it.books
                        if (books != null) {
                            wrap!!.visibility = View.VISIBLE
                            for (i in books.indices) {

                                val writerName = books[i].writerName
                                val subject = books[i].subject
                                val bookImg = books[i].bookImg
                                val isAdult = books[i].isAdult
                                val isFinish = books[i].isFinish
                                val isPremium = books[i].isPremium
                                val isNobless = books[i].isNobless
                                val intro = books[i].intro
                                val isFavorite = books[i].isFavorite
                                val bookCode = books[i].bookCode
                                val categoryKoName = books[i].categoryKoName
                                val cntChapter = books[i].cntChapter

                                items.add(
                                    BookListDataC(
                                        writerName,
                                        subject,
                                        bookImg,
                                        isAdult,
                                        isFinish,
                                        isPremium,
                                        isNobless,
                                        intro,
                                        isFavorite,
                                        bookCode,
                                        categoryKoName,
                                        cntChapter,
                                    )
                                )
                            }
                        } else {
                            blank!!.visibility = View.VISIBLE
                        }
                    }
                    adapter!!.notifyDataSetChanged()
                    if(page == 1){
                        recyclerView!!.layoutManager = linearLayoutManager
                        recyclerView!!.adapter = adapter
                    }
                }
            }

            override fun onFailure(call: Call<BookListResultC?>, t: Throwable) {
                Log.d("onFailure", "실패")
            }
        })

    }

    private var recyclerViewScroll: RecyclerView.OnScrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
            if(!recyclerView.canScrollVertically(1)) {
                cover!!.visibility = View.VISIBLE
                page++
                newBookList(page)
            }

        }
    }

    companion object {

        private const val ARG_SECTION_NUMBER = "section_number"

        fun newInstance(index: Int): NewTab {
            val fragment = NewTab()
            val bundle = Bundle()
            bundle.putInt(ARG_SECTION_NUMBER, index)
            fragment.arguments = bundle
            return fragment
        }
    }
}